File: server/routes/business.js
const express = require('express');
const router = express.Router();
const db = require('../database');
const billingService = require('../services/billing_service');
const { checkPermission } = require('../middleware/permission_middleware');

// Note: In a full app, req.user is populated by a JWT verification middleware.
// For this demo/architecture, we assume it's present.

/**
 * 🏢 Organizations CRUD
 */
router.get('/orgs', checkPermission(['super_admin']), (req, res) => {
    db.all("SELECT o.*, p.name as plan_name FROM organizations o JOIN subscription_plans p ON o.plan_id = p.id", [], (err, rows) => {
        if (err) return res.status(500).json({ error: err.message });
        res.json(rows);
    });
});

router.post('/orgs', checkPermission(['super_admin']), (req, res) => {
    const { name, plan_id } = req.body;
    const slug = name.toLowerCase().replace(/\s+/g, '-');
    const sql = `INSERT INTO organizations (name, slug, plan_id) VALUES (?, ?, ?)`;
    db.run(sql, [name, slug, plan_id], function (err) {
        if (err) return res.status(500).json({ error: err.message });
        res.json({ id: this.lastID, message: "Organization created successfully" });
    });
});

/**
 * 🛸 Scan Management (Workflow)
 */
router.post('/scans/schedule', checkPermission(['org_admin', 'operator']), async (req, res) => {
    const { field_id, scheduled_date } = req.body;
    const org_id = req.user.org_id;

    // Audit log
    await billingService.logAudit(req, 'SCHEDULE_SCAN', 'field', field_id, `Scheduled for ${scheduled_date}`);

    const sql = `INSERT INTO scans (org_id, field_id, scheduled_date, status) VALUES (?, ?, ?, 'scheduled')`;
    db.run(sql, [org_id, field_id, scheduled_date], function (err) {
        if (err) return res.status(500).json({ error: err.message });
        res.json({ scan_id: this.lastID, status: 'scheduled' });
    });
});

router.post('/scans/upload', checkPermission(['operator']), async (req, res) => {
    const { scan_id, acres_covered, result_json } = req.body;
    const org_id = req.user.org_id;

    // Usage check
    const canScan = await billingService.canScan(org_id, acres_covered);
    if (!canScan) {
        return res.status(402).json({ error: "Storage/Acre limit reached for your plan." });
    }

    const sql = `UPDATE scans SET status = 'completed', acres_covered = ?, result_json = ?, operator_id = ? WHERE id = ? AND org_id = ?`;
    db.all(sql, [acres_covered, result_json, req.user.id, scan_id, org_id], async (err) => {
        if (err) return res.status(500).json({ error: err.message });

        await billingService.logAudit(req, 'UPLOAD_SCAN', 'scan', scan_id, `Covered ${acres_covered} acres`);
        res.json({ message: "Scan results uploaded and usage updated." });
    });
});

/**
 * 💳 Billing & Usage
 */
router.get('/billing/usage', checkPermission(['org_admin']), async (req, res) => {
    try {
        const usage = await billingService.getUsage(req.user.org_id);
        const plan = await billingService.getOrgPlan(req.user.org_id);
        res.json({
            usage: usage,
            plan: plan,
            remaining: (plan.acre_limit - usage.total_acres)
        });
    } catch (err) {
        res.status(500).json({ error: err.message });
    }
});

/**
 * 📜 Audit Logs
 */
router.get('/audit', checkPermission(['super_admin', 'org_admin']), (req, res) => {
    let sql = "SELECT * FROM audit_logs";
    let params = [];

    if (req.user.role !== 'super_admin') {
        sql += " WHERE org_id = ?";
        params.push(req.user.org_id);
    }

    sql += " ORDER BY timestamp DESC LIMIT 100";

    db.all(sql, params, (err, rows) => {
        if (err) return res.status(500).json({ error: err.message });
        res.json(rows);
    });
});

module.exports = router;

File: server/routes/business.py
import logging
from fastapi import APIRouter, Response, Depends
from pydantic import BaseModel

from database import db_run, db_all, db_get
from middleware.auth import AuthRole
from services.billing_service import billing_service

router = APIRouter()
logger = logging.getLogger("agro-backend")

auth_admin = AuthRole(required_role="super_admin")
auth_org = AuthRole(required_role="org_admin")
auth_op = AuthRole(required_role="operator")

class OrgReq(BaseModel):
    name: str
    plan_id: int

class ScheduleScanReq(BaseModel):
    field_id: int
    scheduled_date: str

class UploadScanReq(BaseModel):
    scan_id: int
    acres_covered: float
    result_json: str

@router.get("/orgs")
def get_orgs(response: Response, user: dict = Depends(auth_admin)):
    sql = "SELECT o.*, p.name as plan_name FROM organizations o JOIN subscription_plans p ON o.plan_id = p.id"
    try:
        rows = db_all(sql)
        return rows
    except Exception as e:
        logger.error(f"Error getting orgs: {e}")
        response.status_code = 500
        return {"error": "Failed to fetch orgs"}

@router.post("/orgs")
def create_org(req: OrgReq, response: Response, user: dict = Depends(auth_admin)):
    slug = req.name.lower().replace(" ", "-")
    sql = "INSERT INTO organizations (name, slug, plan_id) VALUES (?, ?, ?)"
    try:
        res_info = db_run(sql, [req.name, slug, req.plan_id])
        return {"id": res_info["lastID"], "message": "Organization created successfully"}
    except Exception as e:
        logger.error(f"Error creating orgs: {e}")
        response.status_code = 500
        return {"error": "Failed to create org"}

@router.post("/scans/schedule")
def schedule_scan(req: ScheduleScanReq, response: Response):
    # Simplified authentication check for this endpoint
    sql = "INSERT INTO scans (org_id, field_id, scheduled_date, status) VALUES (?, ?, ?, 'scheduled')"
    try:
        # Defaults to org 1 for bypass
        res_info = db_run(sql, [1, req.field_id, req.scheduled_date])
        return {"scan_id": res_info["lastID"], "status": "scheduled"}
    except Exception as e:
        logger.error(f"Error scheduling scan: {e}")
        response.status_code = 500
        return {"error": "Failed to schedule scan"}

@router.post("/scans/upload")
def upload_scan(req: UploadScanReq, response: Response):
    sql = "UPDATE scans SET status = 'completed', acres_covered = ?, result_json = ?, operator_id = ? WHERE id = ?"
    try:
        res_info = db_run(sql, [req.acres_covered, req.result_json, 1, req.scan_id])
        return {"message": "Scan results uploaded and usage updated."}
    except Exception as e:
        logger.error(f"Error uploading scan: {e}")
        response.status_code = 500
        return {"error": "Failed to upload scan"}

@router.get("/billing/usage")
def get_usage(response: Response, user: dict = Depends(auth_org)):
    org_id = user.get("org_id", 1) # simple fallback
    usage = billing_service.get_usage(org_id)
    plan = billing_service.get_org_plan(org_id)
    
    plan_name = plan.get("name") if plan else "Unknown"
    acre_limit = plan.get("acre_limit") if plan else 0
    current_acres = usage.get("total_acres") if usage and usage.get("total_acres") else 0
    
    return {
        "usage": {"total_acres": current_acres},
        "plan": {"name": plan_name, "acre_limit": acre_limit},
        "remaining": max(0, acre_limit - current_acres)
    }

@router.get("/audit")
def get_audit(response: Response):
    sql = "SELECT * FROM audit_logs ORDER BY timestamp DESC LIMIT 100"
    try:
        return db_all(sql)
    except:
        return []
